### PR TITLE
Add EIP-6123: Smart Derivative Contract

### PR DESCRIPTION
**Pull request by pekola**
For the concept of a smartderivativecontract please find the following:
1. eip markdown file
2. assets/contracts: interface and ref implementation
3. assets/test: unit tests

### CODE REVIEW

File: EIPS/eip-0000.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
+---
+eip: 0000
```
Comment (Pandapip1) [inline EIPS/eip-0000.md:2]:
```suggestion
eip: 6123
```
Make sure to rename the file, too!

File: EIPS/eip-6123.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+### Trade Events
+
+The following events are emitted during a SDC trade livecycle.
+
+#### TradeIncepted
+
+Emitted on trade inception - method 'inceptTrade'
+
+``` js
```
Comment (Pandapip1) [inline EIPS/eip-6123.md:114]:
```suggestion
```solidity
```

File: EIPS/eip-6123.md [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+event TradeIncepted(address initiator, string tradeId, string tradeData);
+```
+
+#### TradeConfirmed
+
+Emitted on trade confirmation - method 'confirmTrade'
+
+``` js
```
Comment (Pandapip1) [inline EIPS/eip-6123.md:122]:
```suggestion
```solidity
```

File: EIPS/eip-6123.md [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+event TradeConfirmed(address confirmer, string tradeId);
+```
+
+#### TradeActivated
+
+Emitted when trade is activated
+
+``` js
```
Comment (Pandapip1) [inline EIPS/eip-6123.md:130]:
```suggestion
```solidity
```

File: EIPS/eip-6123.md [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+event TradeActivated(string tradeId);
+```
+
+#### TradeTerminationRequest
+
+Emitted when termination request is initiated by a counterparty
+
+``` js
```
Comment (Pandapip1) [inline EIPS/eip-6123.md:138]:
```suggestion
```solidity
```

File: EIPS/eip-6123.md [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+event TradeTerminationRequest(address cpAddress, string tradeId);
+```
+
+#### TradeTerminationConfirmed
+
+Emitted when termination request is confirmed by a counterparty
+
+``` js
```
Comment (Pandapip1) [inline EIPS/eip-6123.md:146]:
```suggestion
```solidity
```

File: EIPS/eip-6123.md [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+event TradeTerminationConfirmed(address cpAddress, string tradeId);
+```
+
+#### TradeTerminated
+
+Emitted when trade is terminated
+
+``` js
```
Comment (Pandapip1) [inline EIPS/eip-6123.md:154]:
```suggestion
```solidity
```

File: EIPS/eip-6123.md [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+
+### Process Events
+
+The following events are emitted during SDC's process livecycle.
+
+#### ProcessAwaitingFunding
+
+Emitted when funding phase is initiated
+
+``` js
```
Comment (Pandapip1) [inline EIPS/eip-6123.md:166]:
```suggestion
```solidity
```

File: EIPS/eip-6123.md [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+event ProcessAwaitingFunding();
+```
+
+#### ProcessFunded
+
+Emitted when funding has completed successfully - method 'initiatePrefunding'
+
+``` js
```
Comment (Pandapip1) [inline EIPS/eip-6123.md:174]:
```suggestion
```solidity
```

File: EIPS/eip-6123.md [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+event ProcessFunded();
+```
+
+#### ProcessSettlementRequest
+
+Emitted when a settlement is initiated - method 'initiateSettlement'
+
+``` js
```
Comment (Pandapip1) [inline EIPS/eip-6123.md:182]:
```suggestion
```solidity
```

File: EIPS/eip-6123.md [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+event ProcessSettlementRequest(string tradeData, string lastSettlementData);
+```
+
+#### ProcessSettled
+
+Emitted when settlement was processed successfully - method 'performSettlement'
+
+``` js
```
Comment (Pandapip1) [inline EIPS/eip-6123.md:190]:
```suggestion
```solidity
```

File: EIPS/eip-6123.md [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+
+Method enables an eligible party to request a mutual termination.
+
+``` js
+function requestTradeTermination(string memory tradeId) external;
+```
+
+#### confirmTradeTermination
+
+With this method eligible party is allowed to confirm a former requested mutual trade termination
```
Comment (Pandapip1) [inline EIPS/eip-6123.md:100]:
```suggestion
Allows eligible parties to confirm a formerly-requested mutual trade termination.
```

File: EIPS/eip-6123.md [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+Method serves as callback called from an external oracle providing settlement amount and used settlement data which also get stored.
+Settlement amount will be checked according to contract terms resulting in either a reqular settlement or a termination of the trade.
+
+``` js
+function performSettlement(int256 settlementAmount, string memory settlementData) external;
+```
+
+#### requestTermination
+
+Method enables an eligible party to request a mutual termination.
```
Comment (Pandapip1) [inline EIPS/eip-6123.md:92]:
```suggestion
Allows an eligible party to request a mutual termination
```

File: EIPS/eip-6123.md [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+
+This method checks whether contractual prefunding is provided by both counterparties as agreed in the contract terms. Triggers a contract termination if not. 
+
+``` js
+function initiatePrefunding() external;
+```
+
+#### initiateSettlement
+
+Allows eligible particiants - e.g. counterparties or a delegated agent - to initiate a settlement.
```
Comment (Pandapip1) [inline EIPS/eip-6123.md:74]:
```suggestion
Allows eligible participants (such as counterparties or a delegated agent) to initiate a settlement.
```

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
The interface is a little underdefined. Would you mind adding clarification on which parameters correspond to what (yes, it's pretty obvious, but needs to be specified).

**Comment by pekola** [comment]
> The interface is a little underdefined. Would you mind adding clarification on which parameters correspond to what (yes, it's pretty obvious, but needs to be specified).

Hi - this is not really clear to me - should we comment more on the metods and events and their arguments?

**Comment by pekola** [comment]
> > The interface is a little underdefined. Would you mind adding clarification on which parameters correspond to what (yes, it's pretty obvious, but needs to be specified).
> 
> Hi - this is not really clear to me - should we comment more on the methods and events and their arguments in the interface file?

--> try to do my best to provide more clarification. See ISDC:sol

**Review by cfries — APPROVED** [review]

File: EIPS/eip-6123.md [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+
+![image info](../assets/eip-0000/doc/sdc_trade_and_process_states.png)
+
+### Sequence diagram of trade initiation and settlement livecycle
+
+![image info](../assets/eip-0000/doc/sdc_livecycle_sequence_diagram.png)
+
+## Test Cases
+
+Live-cycle unit tests based on the sample implementation and usage of erc20 token is provided. See folder /assets/test.
```
[the remark this answers is not in the document]
Comment (cfries) [reply EIPS/eip-6123.md]: I have update the reference to the unit tests. All test are in one file (SDC.js) and the eip now references that file.

File: EIPS/eip-6123.md [hunk 16]
```diff
@@ -0,0 +1,244 @@
+---
+eip: 6123
+title: Smart Derivative Contract
+description: A deterministic protocol for frictionless post-trade processing of OTC financial contracts
+author: Christian Fries (@cfries), Peter Kohl-Landgraf (@pekola), Alexandros Korpis (@kourouta)
+discussions-to: https://ethereum-magicians.org/
```
[the remark this answers is not in the document]
Comment (cfries) [reply EIPS/eip-6123.md]: Link to _thread on ethereum magicians_ was added by @pekola

**Review by cfries — APPROVED** [review]

**Review by Pandapip1 — APPROVED** [review]
I think this is fine for a draft

**Comment by cfries** [comment]
> I think this is fine for a draft

Thank you. So it will be merged autmatically? Or anything to do for us? @eth-bot ?

**Review by sackodiego — APPROVED** [review]

**Pull request merged**